src/middleware.js

```diff
@@ -15,6 +14,160 @@ import { handleCarBlock } from './handlers/car-block.js'
+  // TODO: we should implement an SWR pattern here - record an expiry in the metadata and if the expiry has passed, re-validate the cache after
+  // returning the value
+  if (cachedValue) {
+    return deserializeTokenMetadata(cachedValue)
+  } else {
+    const accounting = Accounting.create({ serviceURL: env.ACCOUNTING_SERVICE_URL })
+    const tokenMetadata = await accounting.getTokenMetadata(authToken)
+    if (tokenMetadata) {
+      await env.AUTH_TOKEN_METADATA.put(authToken, serializeTokenMetadata(tokenMetadata))
+      return tokenMetadata
+    } else {
+      return null
+    }
+  }
+}
+
+/**
+ * @type {import('./bindings.js').RateLimits}
+ */
+const RateLimits = {
+  create: ({ env }) => ({
+    check: async (cid, options) => {
+      const authToken = await getAuthorizationTokenFromRequest(options)
+      if (authToken) {
+        console.log(`found token ${authToken}, looking for content commitment`)
+        const tokenMetadata = await getTokenMetadata(env, authToken)
+
+        if (tokenMetadata) {
+          if (tokenMetadata.invalid) {
+            // this means we know about the token and we know it's invalid, so we should just use the CID rate limit
+            return checkRateLimitForCID(env, cid)
+          } else {
+            // TODO at some point we should enforce user configurable rate limits and origin matching
+            // but for now we just serve all valid token requests
+            return RATE_LIMIT_EXCEEDED.NO
+          }
+        } else {
+          // we didn't get any metadata - for now just use the top level rate limit
+          // this means token based requests will be subject to normal rate limits until the data propagates
+          return checkRateLimitForCID(env, cid)
+        }
+      } else {
+        // no token, use normal rate limit
+        return checkRateLimitForCID(env, cid)
+      }
+    }
+  })
+}
+
+/**
+ * @type {import('./bindings.js').Accounting}
+ */
+const Accounting = {
+  create: ({ serviceURL }) => ({
+    record: async (cid, options) => {
+      console.log(`using ${serviceURL} to record a GET for ${cid} with options`, options)
+    },
+
+    getTokenMetadata: async () => {
+      // TODO I think this needs to check the content claims service (?) for any claims relevant to this token
+      // TODO do we have a plan for this? need to ask Hannah if the indexing service covers this?
+      return null
+    }
+  })
+}
+
+/**
+ *
+ * @param {Pick<Request, 'headers'>} request
+ * @returns string
+ */
+async function getAuthorizationTokenFromRequest (request) {
+  // TODO this is probably wrong
+  const authToken = request.headers.get('Authorization')
+  return authToken
+}
+
+/**
+ *
+ * @type {import('@web3-storage/gateway-lib').Middleware<IpfsUrlContext, IpfsUrlContext, Environment>}
+ */
+export function withRateLimits (handler) {
+  return async (request, env, ctx) => {
+    if (!env.FF_RATE_LIMITER_ENABLED) {
```

Comment:
```suggestion
    if (env.FF_RATE_LIMITER_ENABLED !== 'true') {
```

It'll be a string or undefined. Best to explicitly require the string "true".

Reply:
It is typed as a boolean in the Environment bindings.d.ts

Reply:
Hmm, yes I see but that doesn't automatically turn it into a boolean - it's just an environment variable right, either set in `wrangler.toml` or as a deployment variable - it will be a string unless you explicitly cast it in code.

Reply:
Oh ok. That makes sense. Thanks for the clarification.